File: DSA/src/Questions/LinkedList/PartitionList.java
package Questions.LinkedList;
/*
https://leetcode.com/problems/partition-list/description/
 Given the head of a linked list and a value x, partition it such that all nodes less than x come before nodes greater than or equal to x.

You should preserve the original relative order of the nodes in each of the two partitions.

 

Example 1:


Input: head = [1,4,3,2,5,2], x = 3
Output: [1,2,2,4,3,5]
Example 2:

Input: head = [2,1], x = 2
Output: [1,2]
 

Constraints:

The number of nodes in the list is in the range [0, 200].
-100 <= Node.val <= 100
-200 <= x <= 200
Time Complexity: O(N)
Space Complexity: O(1)
 */
public class PartitionList {
    public Node partition(Node head, int x) {
        Node lessHead = new Node(0); // Dummy head for the 'less than x' list
        Node greaterHead = new Node(0); // Dummy head for the 'greater than or equal to x' list
        Node less = lessHead; // Pointer for the 'less than x' list
        Node greater = greaterHead; // Pointer for the 'greater than or equal to x' list

        while (head != null) {
            if (head.val < x) {
                less.next = head; // Add to 'less' list
                less = less.next;
            } else {
                greater.next = head; // Add to 'greater' list
                greater = greater.next;
            }
            head = head.next;
        }

        greater.next = null; // End the 'greater' list
        less.next = greaterHead.next; // Concatenate 'less' list with 'greater' list

        return lessHead.next; // Return the concatenated list, skipping the dummy head

    }
    public static void main(String[] args) {
        
    }
}

File: DSA/Python/Questions/LinkedList/PartitionList.py
"""
# Question: Partition List
# Link: https://leetcode.com/problems/partition-list/

# Partition linked list around value x, maintaining relative order

# Time Complexity: O(n) where n is length of list
# Space Complexity: O(1)

# Algorithm:
# 1. Create two dummy heads for less and greater lists
# 2. Traverse original list and distribute nodes
# 3. Connect less list to greater list
# 4. Return merged result

# Key Components:
# - Node class for linked list structure
# - partition(): Main partitioning implementation
# - Two-list approach with dummy heads
"""


class Node:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class PartitionList:
    def partition(self, head: Node, x: int) -> Node:
        # Create dummy heads
        less_head = Node(0)
        greater_head = Node(0)
        less = less_head
        greater = greater_head

        # Partition nodes
        while head:
            if head.val < x:
                less.next = head
                less = less.next
            else:
                greater.next = head
                greater = greater.next
            head = head.next

        # Connect partitions
        greater.next = None
        less.next = greater_head.next

        return less_head.next


def main():
    # Create test list: 1->4->3->2->5->2, x = 3
    head = Node(1)
    head.next = Node(4)
    head.next.next = Node(3)
    head.next.next.next = Node(2)
    head.next.next.next.next = Node(5)
    head.next.next.next.next.next = Node(2)

    solution = PartitionList()
    result = solution.partition(head, 3)

    # Print result
    while result:
        print(result.val, end=" ")
        result = result.next


if __name__ == "__main__":
    main()
